buildkite/bazel_auto_sheriff.py

```diff
@@ -0,0 +1,643 @@
+        # Be smart here to reduce log length.
+        if self.analyze_log[-1].startswith(COLORS["INFO"] + "Waiting for "):
+            self.analyze_log.pop(-1)
+        self.__log("INFO", text)
+
+
+    def __trigger_bisect(self, tasks):
+        env = {
+            "PROJECT_NAME": self.project,
+            "TASK_NAME_LIST": ",".join(tasks) if tasks else "",
+        }
+        return CULPRIT_FINDER_PIPELINE_CLIENT.trigger_new_build("HEAD", f"Bisecting {self.project}", env)
+
+
+    # Search for certain message in the log to determine the bisect result.
+    # Return values are
+    #   1. A bisect result message
+    #   2. The commit of the culprit if found, otherwise None
+    def __determine_bisect_result(self, job):
+        bisect_log = CULPRIT_FINDER_PIPELINE_CLIENT.get_build_log(job)
+        pos = bisect_log.rfind("first bad commit is ")
+        if pos != -1:
+            start = pos + len("first bad commit is ")
+            culprit_commit = bisect_log[start:start + 40]
+            return "\n".join([
+                 "Culprit found!",
+                 bisect_log[pos:].replace("\r", ""),
+                 "Bisect URL: " + job["web_url"],
+            ]), culprit_commit
+        pos = bisect_log.rfind("Given good commit")  # Matching "Given good commit (XXXX) is not actually good, abort bisecting."
+        if pos != -1:
+            self.broken_by_infra = True
+            return "\n".join([
+                "Given good commit is now failing. This is probably caused by remote cache issue or infra change.",
+                "Please ping philwo@ or pcloudy@ for investigation.",
+                "Bisect URL: " + job["web_url"],
+            ]), None
+        pos = bisect_log.rfind("first bad commit not found, every commit succeeded.")
+        if pos != -1:
+            return "\n".join([
+                "Bisect didn't manage to reproduce the failure, all builds succeeded.",
+                "Maybe the build are cached from previous build with a different Bazel version or it could be flaky.",
+                "Please try to rerun the bisect with NEEDS_CLEAN=1 and REPEAT_TIMES=3.",
+                "Bisect URL: " + job["web_url"],
+            ]), None
+        return "Bisect failed due to unknown reason, please check " + job["web_url"], None
+
+
+    def __retry_failed_jobs(self, build_result, buildkite_client):
+        retry_per_failed_task = {}
+        for task, info in build_result["tasks"].items():
+            if info["state"] != "passed":
+                retry_per_failed_task[task] = buildkite_client.trigger_job_retry(build_result["build_number"], info["id"])
+        for task, job_info in retry_per_failed_task.items():
+            retry_per_failed_task[task] = buildkite_client.wait_job_to_finish(build_number = build_result["build_number"], job_id = job_info["id"], logger = self)
+        return retry_per_failed_task
+
+
+    def __print_job_list(self, jobs):
+        for job in jobs:
+            name = job["name"]
```

Comment:
no need for these variables - you can access dicts in the f string

Reply:
Ah... I thought I couldn't